File: src/net/azib/photos/AlbumLoader.java
package net.azib.photos;

import java.util.ArrayList;

import static java.lang.Float.parseFloat;
import static java.lang.Integer.parseInt;
import static java.lang.Long.parseLong;

public class AlbumLoader implements XMLListener<Album> {
  private Album album;
  private Photo photo;
  private Comment comment;

  @Override public Album getResult() {
    return album;
  }

  @Override
  public void rootElement(String name) {
    album = new Album();
    photo = null;
    comment = null;
  }

  @Override
  public void rootElementEnd(String name) {
  }

  @Override
  public void value(String path, String value) throws StopParse {
    switch (path) {
      case "name": album.setName(value); return;
      case "title": album.setTitle(value); return;
      case "subtitle": album.setDescription(value); return;
      case "icon": album.setThumbUrl(value); return;
      case "timestamp": album.setTimestamp(parseLong(value)); return;
      case "nickname": album.setAuthor(value); return;
      case "user": album.setAuthorId(value); return;
      case "access": album.setPublic("public".equals(value)); return;
      case "numphotos": album.setSize(parseInt(value));
                        album.setPhotos(new ArrayList<Photo>(album.getSize())); return;
      case "where/Point/pos": album.setGeo(new GeoLocation(value)); return;

      case "entry/category@term":
        if (value.endsWith("photo")) album.getPhotos().add(photo = new Photo());
        else if (value.endsWith("comment")) album.getComments().add(comment = new Comment());
        return;
    }

    if (photo != null) {
      switch (path) {
        case "entry/id": photo.setId(value); break;
        case "entry/title": photo.setTitle(value); break;
        case "entry/summary": photo.setDescription(value); break;
        case "entry/timestamp": photo.setTimestamp(parseLong(value)); break;
        case "entry/width": photo.setWidth(parseInt(value)); break;
        case "entry/height": photo.setHeight(parseInt(value)); break;
        case "entry/content@src": photo.setUrl(value); break;
        case "entry/group/thumbnail@url": photo.setThumbUrl(value); break;
        case "entry/tags/fstop": photo.getExif().fstop = parseFloat(value); break;
        case "entry/tags/exposure": photo.getExif().exposure = parseFloat(value); break;
        case "entry/tags/focallength": photo.getExif().focal = parseFloat(value); break;
        case "entry/tags/iso": photo.getExif().iso = value; break;
        case "entry/tags/model": photo.getExif().camera = value; break;
        case "entry/where/Point/pos": photo.setGeo(new GeoLocation(value)); break;
      }
    }
    else if (comment != null) {
      switch (path) {
        case "entry/content": comment.text = value; break;
        case "entry/author/name": comment.author = value; break;
        case "entry/author/thumbnail": comment.avatarUrl = value; break;
        case "entry/author/user": comment.authorId = value; break;
        case "entry/photoid": comment.photoId = value; break;
      }
    }
  }

  @Override
  public void start(String path) throws StopParse {
  }

  @Override
  public void end(String path) throws StopParse {
    if ("entry".equals(path)) {
      photo = null;
      comment = null;
    }
  }
}

File: src/net/azib/photos/AlbumLoader.kt
package net.azib.photos

class AlbumLoader(thumbSize: Int) : XMLListener<Album> {
  private var photo: Photo? = null
  private var comment: Comment? = null
  override val result = Album(thumbSize)

  override fun value(path: String, value: String) {
    result.apply {
      when (path) {
        "id" -> id = value
        "name" -> name = value
        "title" -> title = value
        "subtitle" -> description = value
        "icon" -> thumbUrl = value + ".jpg"
        "timestamp" -> timestamp = value.toLong()
        "nickname" -> author = value
        "user" -> authorId = value
        "access" -> isPublic = "public" == value
        "numphotos" -> {
          size = value.toInt()
          photos.ensureCapacity(size)
        }
        "where/Point/pos" -> geo = GeoLocation(value)
        "entry/category@term" -> when {
          value.endsWith("photo") -> {
            photo = Photo()
            photos.add(photo!!)
          }
          value.endsWith("comment") -> {
            comment = Comment()
            comments.add(comment!!)
          }
        }
      }
    }

    photo?.apply { 
      when (path) {
        "entry/id" -> id = value
        "entry/title" -> title = value
        "entry/summary" -> description = value
        "entry/timestamp" -> timestamp = value.toLong()
        "entry/width" -> width = value.toInt()
        "entry/height" -> height = value.toInt()
        "entry/content@src" -> url = replaceUrlSuffix(value)
        "entry/group/thumbnail@url" -> thumbUrl = replaceUrlSuffix(value)
        "entry/tags/fstop" -> exif.fstop = value.toFloat()
        "entry/tags/exposure" -> exif.exposure = value.toFloat()
        "entry/tags/focallength" -> exif.focal = value.toFloat()
        "entry/tags/iso" -> exif.iso = value
        "entry/tags/model" -> exif.camera = value
        "entry/where/Point/pos" -> geo = GeoLocation(value)
      }
    }
    
    comment?.apply {
      when (path) {
        "entry/content" -> text = value
        "entry/author/name" -> author = value
        "entry/author/thumbnail" -> avatarUrl = value
        "entry/author/user" -> authorId = value
        "entry/photoid" -> photoId = value
      }
    }
  }

  private val invalidUrlChars = "[?&!%.,:;+*/()'\"]".toRegex()
  private fun replaceUrlSuffix(url: String): String {
    var desc = photo?.description?.take(30)?.replace(invalidUrlChars, "")?.replace(' ', '-')
    if (desc.isNullOrEmpty()) desc = result.photos.size.toString()
    return url.substring(0, url.lastIndexOf('/') + 1) + "${result.name}-${desc}.jpg"
  }

  override fun end(path: String) {
    if ("entry" == path) {
      photo = null
      comment = null
    }
  }
}
